Support for multiarm ancova

gowerc opened the pull request:
Closes #145 

Hey @danielinteractive  /  @tobiasmuetze,

This contains my initial code proposal for implementing multiarm ancova support (note that it is currently missing documentation and tests).

Some key points:

- To preserve backwards compatibility I've left the original `ancova` function unchanged. To support multiple groups a new naming scheme is required for the parameters which would break any existing code which is looking for the current names like `trt` / `lsm_rf`.

- That is, if users want to use the new multi-group ancova they will have to explicitly set `fun=ancova_m_groups` in the call to `analyse()`

- For the new naming convention I've just used the level numbers e.g.   `lsm_L0` = least squared means for the reference, `lsm_L1` = least squared means for the first offset / coefficient.  I've also used a more explicit `trt_L1_L0` to mean the coefficient of the first offset from the reference.   

- As I think we agreed I'm currently only extracting the treatment effects against the reference e.g.  `L1 - L0` & `L2 - L0`. I'm not calculating / extracting all pairwise combinations. 

- For the naming convention I didn't want to use the user-provided group/level values as that can lead to a ton of special edge cases with whitespaces / special characters that are hard to predict and make value-extraction very error prone. Instead I opted for just renaming the group variable to `rbmiGroup` and re-leveling it to be `L0, L1, etc`. Only complication here was that the formula provided in theory can have interaction terms so had to create a function to recursively traverse the AST updating any references of `group` to `rbmiGroup`. I will need to put in extensive unit testing to make sure theres no edge cases here with things like `I()` functions. 

If there any no main objections here I'll start trying to add docs and tests.

gowerc committed:
initial

gowerc committed:
improve formula find and replace

gowerc committed:
added tests for frm_find_and_replace

gowerc committed:
added rest of frm_find_and_replace tests

gowerc committed:
all hail lintr

gowerc committed:
run roxygen

gowerc committed:
added ancova unit tests

gowerc committed:
fix examples for non-exported function

gowerc committed:
updated news.md file

gowerc committed:
added docs

gowerc committed:
fix broken test

gowerc committed:
updated docs / specs

gowerc committed:
updated faq

gowerc committed:
update version number

gowerc committed:
roxygen

gowerc committed:
Merge branch 'main' into 145-multi-ancova

danielinteractive commented:
> Interestingly despite what their documentation says I don't think they actually support more than 2 arms, in particular they have near the top of the function `checkmate::assert_factor(data[[group]], n.levels = 2L)` which would error if there were more than 2 arms.

Hmm... ok weird. But I am pretty sure we wanted this to work with multiple arms, that is also what the other code suggests... I will need to check in September when I am back.

> I am very conflicted here, I agree with everything you said and I am also not happy with splitting it out but that being said I don't see an easy way of merging them without breaking backwards compatibility as the naming scheme has to be different in order for it to make sense we currently have:
> 
> ```
> lsm_alt_visit_1
> lsm_ref_visit_1
> trt_visit_1
> ```

Yeah I understand the need for backwards compatibility. But I wonder if an alternative solution here could be to have this old naming scheme used for 2 arms, and the new naming scheme used for more than 2 arms?

> In hindsight these are already bad names that confuse users. If we were to extend this whilst maintaining backwards compatibility we'd end up with something like:
> 
> ```
> lsm_alt_visit_1
> lsm_alt2_visit_1
> lsm_ref_visit_1
> trt_visit_1
> trt_alt2_visit_1
> ```
> 
> Which I would argue is even worse.

Personally I think that would also be acceptable. The user will not see these names too much anyway if they just use the `rbmi` functions.

> 
> Alternatively we could just have some code that dispatches to the different ANCOVA function based on the number of levels in group variable but this makes the documentation / explanation clunky e.g.

Yeah I would not have two functions necessarily but just two naming schemes. Or just go with the naming scheme just described which I think is fine.

> One option though to minimise maintenance is that we could deprecate `anocva` with a note that it will be removed in the future and replaced with `ancova_m_groups` then in 6 months to a years time we could fully remove it, we could also create an alias of `ancova = ancova_m_groups` after removing so we don't have such an awkward name.
> 
> (Despite my tone I am not confident on the best path forward here as all options appear bad to me so please do challenge if you still disagree)

Personally I would just go with the naming scheme you suggested, I think that is reasonable. Just for the sake of a better naming scheme I would not go the route of two functions or function deprecation. Just my 2 cents 😄

wwojciech commented on R/ancova.R:
```diff
@@ -206,3 +322,110 @@ ancova_single <- function(
+    assert_that(
+        is.factor(data[[group]]),
+        length(levels(data[[group]])) >= 2,
+        length(unique(data[[group]])) == length(levels(data[[group]])),
```
@tobiasmuetze - Can we please remove this from the assertion:
```r
length(unique(data[[group]])) == length(levels(data[[group]]))
```

and then, of course, modify the code below that computes `trt` to:

```r
trt <- lapply(
  levels(data2[["rbmiGroup"]])[-1],
  function(x) {
    term <- sprintf("rbmiGroup%s", x)
    if (term %in% names(coef(mod))) {
      list(
        est = coef(mod)[[term]],
        se = sqrt(vcov(mod)[term, term]),
        df = df.residual(mod)
      )
    } else {
      list(est = NA, se = NA, df = NA)
    }
  }
)

```
This is because there are some cases where we have no records for a certain level of `data[[vars$group]]`, and this should not cause the function to stop.

tobiasmuetze replied:
Is my understanding correct that if this assertion is removed, the group variable in the input data set could have a certain level but no subjects with this level? In which clinical trial scenario would this be relevant?

wwojciech replied:
> Is my understanding correct that if this assertion is removed, the group variable in the input data set could have a certain level but no subjects with this level? In which clinical trial scenario would this be relevant?

Yes, it is. There might be no data for a given level of `data[[vars$group]]`.

This is relevant in scenarios where we fit an ANCOVA model using a subset of the full dataset defined by a subgroup variable (such as geographic region). In that case, for some levels of the subgroup variable, certain levels of `data[[vars$group]]` may be absent from the subset.

For example, in a given geographic region, we may not have data for all treatment groups. See the example table below for a dataset with three treatment groups (i.e., `data[[vars$group]]` has three levels), where the table presents results for only two treatment groups of interest.

<img width="398" height="435" alt="example_table" src="https://github.com/user-attachments/assets/3ebfd5cc-0fab-46e8-afe9-99bf9705925f" />

tobiasmuetze replied:
I’d suggest keeping the assertion, because this scenario seems to point to an inadequate model/data setup rather than something the function should accommodate.

For subgroup analyses based on multiply imputed data, the treatment-by-subgroup interaction should be included in the imputation model. If a treatment-by-subgroup combination is completely absent in the data, then fitting such an imputation model should already be problematic. 

In that sense, the assertion `length(unique(data[[group]])) == length(levels(data[[group]]))` at least ensures that all declared treatment levels are actually represented in the dataset being analyzed. If a level is absent after subsetting, that should probably be treated as a signal that the analysis model is not adequate for the intended subgroup analysis.

The statistical concern is that if the imputation model does not include the treatment-by-subgroup interaction, subsequent subgroup-specific treatment effect estimates, or analyses of treatment-by-subgroup interaction, may be biased toward the null when interactions are present; see [Schafer](https://journals.sagepub.com/doi/10.1177/096228029900800102). More generally, [Austin et al.](https://doi.org/10.1016/j.cjca.2020.11.010) state that “it is important to include in the imputation model all the variables that will be included in the analysis model.”

wwojciech replied:
> I’d suggest keeping the assertion, because this scenario seems to point to an inadequate model/data setup rather than something the function should accommodate.
> 
> For subgroup analyses based on multiply imputed data, the treatment-by-subgroup interaction should be included in the imputation model. If a treatment-by-subgroup combination is completely absent in the data, then fitting such an imputation model should already be problematic.
> 
> In that sense, the assertion `length(unique(data[[group]])) == length(levels(data[[group]]))` at least ensures that all declared treatment levels are actually represented in the dataset being analyzed. If a level is absent after subsetting, that should probably be treated as a signal that the analysis model is not adequate for the intended subgroup analysis.
> 
> The statistical concern is that if the imputation model does not include the treatment-by-subgroup interaction, subsequent subgroup-specific treatment effect estimates, or analyses of treatment-by-subgroup interaction, may be biased toward the null when interactions are present; see [Schafer](https://journals.sagepub.com/doi/10.1177/096228029900800102). More generally, [Austin et al.](https://doi.org/10.1016/j.cjca.2020.11.010) state that “it is important to include in the imputation model all the variables that will be included in the analysis model.”

Thank you @tobiasmuetze 

My understanding is as follows.

Assume that the imputation model was correctly specified (following the recommendations of Schafer and Austin), i.e., the treatment-by-subgroup interaction is included in the imputation model. Even in this case, it may still happen that, for certain levels of the subgroup variable, not all treatment groups are represented in the observed data. As a result, after imputation there will still be no observations for that particular treatment-by-subgroup combination, since multiple imputation imputes missing values for existing subjects rather than creating new observations.

However, I do not think this necessarily prevents a valid analysis within that subgroup. If the estimand of interest does not involve the missing treatment level, then contrasts among the remaining observed treatment groups remain estimable. In that situation, I would expect the ANCOVA model to fit successfully and to return valid estimates for the estimable contrasts, while returning `NA` for treatment effects involving the absent treatment level.

For this reason, I think the assertion may be too restrictive.

@luwidmer, @danielinteractive - what do you think?

luwidmer committed:
Support for multiarm ancova, Draft v2 (#590)

First draft to further iterate on, according to the discussion in #520
so far @gowerc @tobiasmuetze @bailliem @wwojciech @danielinteractive:
this branch builds on the original multi-arm ANCOVA prototype (#520 )
with a more flexible contrast API.

Let me know what you think, this definitely will need further review and
iteration.


### Flexible custom contrasts

- `group_contrasts` now accepts, in addition to pairwise `c(minuend,
subtrahend)` character vectors, general linear contrasts as named
numeric weight vectors over the group levels (e.g. `c(Placebo = -1, A =
0.5, B = 0.5)` for pooled active-vs-control, or dose-response trends).
Weights must sum to zero; evaluation is coding-agnostic and preserves
the `trt` estimand (evaluated at the covariate reference).
- Default two-arm and multi-arm output is unchanged (`trt`, `lsm_ref`,
`lsm_alt`, `trt_alt2`, ...).

### Correctness & robustness

- Contrast coding-agnostic evaluation: `ancova_linear_contrast()` now
maps group-level weights onto the model coefficients via the factor's
own contrast matrix instead of parsing `rbmiGroupL*` coefficient names.
Results are correct under any `options("contrasts")` (e.g. `contr.sum`),
not just `contr.treatment`, and reduce exactly to the previous
coefficient contrast for the pairwise/default case. Rank-deficient
designs (missing/aliased/`NA` group coefficients) now error instead of
silently dropping a term.
- `as_data_frame_internal()` errors if pooled parameters have no
matching metadata rather than emitting silent `NA` columns.

### Named contrasts & output metadata

- Explicit contrasts must be named: the list name becomes the output
`parameter` and a new `contrast_label` metadata column (surfaced by
`pool()` / `as.data.frame()`). The auto-derived `trt_alt2_alt`-style
names are used only for the default (`group_contrasts = NULL`) set,
preserving backward compatibility. Naming a contrast `trt` reproduces
the classic name.
- Names starting with `lsm_` are rejected (reserved for least-squares
means) to prevent output collisions.
- `pool()` / `as.data.frame()` output gains `contrast_label` alongside
the existing `estimate_type`, `group`, `group_level_1`, `group_level_2`,
`visit` columns; the legacy `parameter` column is retained.

## Documentation

- Fixed stale `ancova()` `@return` (now documents the extended default
scheme and named custom contrasts); added an `@details` note clarifying
that `trt` is the model-coefficient contrast and differs from `lsm_alt −
lsm_ref` under group×covariate interactions.
- Added runnable multi-arm/custom-contrast examples to `ancova()` and
worked pairwise + weight-vector blocks to the FAQ and quickstart
vignettes.
- NEWS updated.

### Tests

- Coefficient-contrast reproduction, pooled/weight-vector correctness,
`contr.sum` invariance, named-label propagation through `pool()`,
reserved-name and validation errors, an end-to-end `ancova()` → `pool()`
→ `as.data.frame()` metadata integration test.

### Housekeeping

- Project formatted with `air`.

### Compatibility

- Two-arm and default multi-arm results and names are unchanged.
`group_contrasts` and the metadata columns are new (unreleased), so
tightening explicit contrasts to require names introduces no break for
released behaviour.
- The covariate reference point for `trt` is fixed at 0 to preserve the
historic estimand (documented in `@details`).

---------

Signed-off-by: Lukas Widmer <l.widmer@gmail.com>
Co-authored-by: Isaac Gravestock <83659704+gravesti@users.noreply.github.com>
Co-authored-by: Craig Gower-Page <craiggower@gmail.com>
Co-authored-by: Lukas Widmer <lukas_andreas.widmer@novartis.com>

Commit:
Fix merge conflicts with main

Merge remote-tracking branch 'origin/main' into 145-multi-ancova

# Conflicts:
#	NEWS.md
#	R/ancova.R
#	R/pool.R
#	man/ancova.Rd
#	man/pool.Rd
#	vignettes/FAQ.html
#	vignettes/advanced.html
#	vignettes/quickstart.Rmd
#	vignettes/quickstart.html

munoztd0 commented:
@luwidmer and @tobiasmuetze sorry to bother but any plans to merge it into main now that the new API has been merged in ?

danielinteractive commented:
I understand that we have an ooo situation so need to wait a couple of days - that said, if there is any support needed here happy to jump in any time from my side.

danielinteractive commented on vignettes/quickstart.Rmd:
```diff
@@ -229,6 +229,55 @@ function which determines the names of the key variables within the data and the
 
 Please also note that the names of the analysis estimates contain `ref` and `alt` to refer to the two treatment arms. In particular `ref` refers to the first factor level of `vars$group` which does not necessarily
 coincide with the control arm. In this example, since `levels(dat[[vars$group]]) = c("DRUG", PLACEBO`), the results associated with `ref` correspond to the intervention arm, while those associated with `alt` correspond to the control arm.
+<<<<<<< HEAD
```
seems some conflict markers are still in here

danielinteractive commented on vignettes/quickstart.Rmd:
```diff
@@ -229,6 +229,55 @@ function which determines the names of the key variables within the data and the
+
+See the documentation for `ancova()` and `set_vars()` for more details.
+=======
+>>>>>>> origin/main
```
here too

danielinteractive commented on NEWS.md:
```diff
@@ -1,5 +1,23 @@
+* The `data.frame` produced by `as.data.frame()` / `pool()` now includes explicit
+  `estimate_type`, `group`, `group_level_1`, `group_level_2`, `contrast_label` and `visit`
+  columns when the analysis was performed with `ancova()`. The existing `parameter` column
+  is retained for backwards compatibility. (#520)
```
Maybe combine this into a single item talking about ancova for more than two groups?

danielinteractive reviewed (commented):
Thanks again @luwidmer for the nice proposal, I just had a quick look through and it looks really nice. Also the testing seems to be very thorough which is great to see, especially that we get full backwards compatibility. Minor things below.
